Demand Paging support for ARM64

npitre opened the pull request:
This includes demand paging support for ARM64.

npitre committed:
arm64: mmu: be stricter about free page entries

Let's consider free entries as being completely zeroed. Future patches
for demand paging support will populate entries and still mark them as
"invalid" which should not be considered free. Current code always clear
entries to be freed so no issues there.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
arm64: mmu: architecture specific hooks for demand paging support

A page table entry used for demand paging is always from the last level
page table and never completely zeroed. Anything else is considered
ARCH_DATA_PAGE_NOT_MAPPED.

Loaded pages use a PTE_PAGE_DESC table entry type. Paged-out pages use a
PTE_INVALID_DESC table entry type and the physical address field is
reused to hold the backing store location token value.

ARCH_DATA_PAGE_ACCESSED corresponds to the AF flag. It is initially
unset and manually set to catch when pages are being accessed and
eventually do something about it.

ARCH_DATA_PAGE_DIRTY corresponds to the lack of the AP_RO flag.
Similarly to the AF flag, pages are initially made read-only to catch
writes.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
arm64: mmu: access fault handler for demand paging

This has two purposes: maintain the accessed and dirty page states, or
call the generic demand paging fault handler otherwise.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
arm64: demand paging is supported

Test configs for UP and SMP are also included.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
arm64: demand_paging: add the k_mem_paging_eviction_accessed() call

Needed for the LRU eviction algorithm to be effective.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

dcpleung commented on arch/arm64/core/mmu.c:
```diff
@@ -1330,3 +1338,187 @@ void z_arm64_swap_mem_domains(struct k_thread *incoming)
+	uint64_t *pte = get_pte_location(&kernel_ptables, virt);
+	uint64_t desc, mask;
+
+	mask = GENMASK(47, PAGE_SIZE_SHIFT);
```
Might be a good idea to make 47 into a named constant (e.g. macro).

dcpleung commented:
Hm... I don't see linker script magic for demand paging here. How do the boot and pinned symbols being placed? Because if the fault handling is paged out, the whole thing would fall apart.

npitre commented:
> Hm... I don't see linker script magic for demand paging here. How do 
> the boot and pinned symbols being placed? Because if the fault handling is paged out, the whole thing would 
> fall apart.

At the moment it is relying on the whole Zephyr image being pinned. Only 
anonymous memory is evictable ... to the RAM backing store for now.
I'm still focussing on the basics.

Talking about those `__pinned` tags: I don't think this is scalable nor
convenient. It would be much easier to do it the other way around i.e.
explicitly mark evictable stuff instead.

The target I'm working on has about 8MB of RAM and 16MB of firmware in flash.
The backing store driver would go through flashmap on top of the flash layer
on top of the SPI layer on top of the DMA layer. Oh and possibly a filesystem
too. That's a lot of code to tag and a lot of extra care expected from anyone
modifying such code even if they're not demand-paging users. This means some
piece of code/data might lack a `__pinned` tag and the crash might happen at
some undeterministic moment.

If instead we consider tagging code/data that can be evicted� then it is no
longer critical if some tags are missing. The `__boot` tags follow that model.

What do you think?

dcpleung commented:
The original idea is that `__pinned_*` are used for symbols that must be in memory and also common enough that everyone would need. Anything applications specific can choose to use `__pinned` (for custom code) or linker script (for code in zephyr main repo) to further fine tune to their needs. The assumption is simply that the set of symbols needed to be always in memory is much smaller, and these is less effort to mark them. If you switch to evictable model, one would probably spend more time on marking evictable ones instead. Not to mention that you will need to audit the current code and mark symbols evictable so that the tests can run.

dcpleung commented on tests/kernel/mem_protect/demand_paging/boards/qemu_cortex_a53.conf:
```diff
@@ -0,0 +1,10 @@
+
+CONFIG_SRAM_SIZE=320
+
+CONFIG_USERSPACE=n
```
It would a good idea to comment why userspace is forcibly disabled here.

npitre commented:
> It would a good idea to comment why userspace is forcibly disabled here.

Because I'm still working on it.
With luck I'll be able to push another revision with this restriction removed.

dcpleung commented:
In that case, do you mind converting this into a draft? So that we know you are still working on it and would refrain from tiny comments.

npitre commented:
> In that case, do you mind converting this into a draft? So that we know you are still working on it and would refrain from tiny comments.

Good idea. Didn't know you could do that.

Comments still accepted though.  ;-)

In particular, what should be the semantic with anonymous mem mappings
and user threads?

dcpleung commented:
Anonymous mappings can be opened to all user threads by specifying `K_MEM_PERM_USER` during mapping. Otherwise, access can be granted via memory domains... at least according to the document. I don't think we even have tests for this.

npitre committed:
arm64: demand_paging: add support for on-demand mappings

This makes ARM64 compatible with K_MEM_MAP_UNPAGED.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

cfriedt commented on arch/arm64/core/mmu.c:
```diff
@@ -755,6 +780,12 @@ static void invalidate_tlb_all(void)
 
+static inline void invalidate_tlb_page(uintptr_t virt)
+{
+	/* to be refined */
+	invalidate_tlb_all();
```
Looks good to me - this was the only thing that kind of stood out.

Are the necessary bits for more refined tlb invalidation are not yet available?

npitre committed:
arm64: demand_paging: allow page fault processing with IRQs enabled

Convention is to call k_mem_page_fault() with IRQs enabled if they were
enabled when the fault occurred.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
demand_paging: add a semihosting based backing store for on-demand sections

This is especially handy for tests and validation using QEMU.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
tests: demand_paging: move existing test to a mem_map subdirectory

This test concerns itself mainly with anonymous memory mappings.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
tests: demand_paging: add a test for on-demand sections

Exercises linker placement, the ondemand backing store, demand paging.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

npitre committed:
sample: demand_paging: add a demo about ondemand section usage

This sample demonstrates how demand paging can be leveraged to deal with
firmware bigger than the available RAM if XIP is not possible. Select
code can be tagged to be loaded into memory on demand, and also be
automatically evicted for more code to be executed when memory is
exhausted.

Signed-off-by: Nicolas Pitre <npitre@baylibre.com>

kartben commented on samples/demand_paging/README.rst:
move the sample to `samples/subsys/demand_paging`

kartben commented on samples/demand_paging/README.rst:
```diff
@@ -0,0 +1,49 @@
+.. _hello_world:
+
+Demand Paging
+#############
```
this will help with discoverability by showing the sample there: https://docs.zephyrproject.org/latest/kernel/memory_management/demand_paging.html#api-reference

```suggestion
.. zephyr:code-sample:: demand-paging
   :name: Demand paging
   :relevant-api: mem-demand-paging

   Leverage demand paging to deal with code bigger than available RAM.
```

kartben commented on samples/demand_paging/README.rst:
```diff
@@ -0,0 +1,49 @@
+Requirements
+************
+
+This demo requires the presence of a MMU in hardware that Zephyr supports,
```
nit/grammar

```suggestion
This demo requires the presence of a supported hardware MMU,
```

kartben commented on samples/demand_paging/src/main.c:
```diff
@@ -0,0 +1,77 @@
+	foo_val2 = 45454545;
+
+#define CODE0 foo_val1 *= foo_val2; foo_val2 += 1234567; foo_val1 -= 9876543210;
+#define CODE1 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0 CODE0
+#define CODE2 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1 CODE1
+#define CODE3 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2 CODE2
+#define CODE4 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3 CODE3
+#define CODE5 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4 CODE4
+
+	/* This is 100000 repetitions of CODE0 */
+	CODE5
+
+#undef CODE5
+#undef CODE4
+#undef CODE3
+#undef CODE2
+#undef CODE1
+#undef CODE0
```
not necessarily a request for change but perhaps it would be more readable / self-explanatory like this? sadly, this is not any faster :)

```suggestion
#define CODE_REPEAT_10(x)     x x x x x x x x x x
#define CODE_REPEAT_100(x)    CODE_REPEAT_10(CODE_REPEAT_10(x))
#define CODE_REPEAT_1000(x)   CODE_REPEAT_10(CODE_REPEAT_100(x))
#define CODE_REPEAT_10000(x)  CODE_REPEAT_10(CODE_REPEAT_1000(x))
#define CODE_REPEAT_100000(x) CODE_REPEAT_10(CODE_REPEAT_10000(x))

	CODE_REPEAT_100000(CODE0)

#undef CODE0
```

npitre commented:
Thanks @kartben for your suggestions. I just put this demo together 
before leaving on vacation. Will address your comments (and hopefully 
others) when I'll return.

npitre commented:
Several fixes, rebased on latest main branch, refreshed included unmerged PRs.

dcpleung commented on tests/kernel/mem_protect/demand_paging/testcase.yaml:
```diff
@@ -6,7 +6,9 @@ tests:
       - demand_paging
-    platform_allow: qemu_x86_tiny
+    platform_allow:
+      - qemu_cortex_a53
```
Maybe also add `qemu_corter_a53/smp` as this PR also enables SMP for demand paging.

dcpleung commented on tests/kernel/mem_protect/demand_paging/mem_map/testcase.yaml:
Will also need to change the name of the tests in the yaml file.

dcpleung commented on tests/kernel/mem_protect/demand_paging/ondemand_section/testcase.yaml:
```diff
@@ -0,0 +1,12 @@
+    - mmu
+    - demand_paging
+tests:
+  ondemand.arm64.semihost:
```
I believe that the test name needs to (somehow) follow the directory structure.

dcpleung commented on arch/arm64/core/mmu.c:
```diff
@@ -1351,3 +1360,197 @@ void z_arm64_swap_mem_domains(struct k_thread *incoming)
+		status |= ARCH_DATA_PAGE_DIRTY;
+	}
+
+	if (clear_accessed) {
+		desc &= ~PTE_BLOCK_DESC_AF;
+		*pte = desc;
+		MMU_DEBUG("page_info: virt=%#lx (clearing AF)\n", virt);
+		debug_show_pte(pte, XLAT_LAST_LEVEL);
+		sync_domains(virt, CONFIG_MMU_PAGE_SIZE, "unaccessible");
+		invalidate_tlb_page(virt);
+	}
```
Hm... `clear_accessed` is supposed to only clear the access state (actually the `ARCH_DATA_PAGE_ACCESSED` state, and associated hardware bits) and not about making next access to generate any exceptions. If NRU is used, there is no need to invalidate the TLB but only needs the access bits cleared.

So maybe guard the TLB invalidation with #ifdef?

npitre commented:
> Hm... `clear_accessed` is supposed to only clear the access state (actually the `ARCH_DATA_PAGE_ACCESSED` state, and associated hardware bits) and not about making next access to generate any exceptions. If NRU is used, there is no need to invalidate the TLB but only needs the access bits cleared.

The accessed state is not managed by the hardware. If the tlb is not 
cleared, further accesses might never be noticed. Only setting the 
access flag doesn't need a tlb clear as without AF no tlb load takes 
place.




> 
> So maybe guard the TLB invalidation with #ifdef?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/zephyrproject-rtos/zephyr/pull/74129#pullrequestreview-2273535604
> You are receiving this because you authored the thread.
> 
> Message ID: ***@***.***>

npitre commented:
- Rebased to purge already merged commits

- Reworked demand-paging sample to get rid of the extreme compilation time

- Addressed latest comments

I think this is ready for merge consideration now.

npitre commented:
Contains only ARM64 specific code now.
Ready for merge consideration. Please review.
